src/main/java/org/apache/ibatis/executor/keygen/Jdbc3KeyGenerator.java

```diff
@@ -92,25 +89,83 @@ public void processBatch(MappedStatement ms, Statement stmt, Collection<Object>
+              + "Specified key properties are " + ArrayUtil.toString(keyProperties) + " and available parameters are "
+              + paramMap.keySet());
+    }
+    String paramName = keyProperties[0].substring(0, firstDot);
+    Object param;
+    if (paramMap.containsKey(paramName)) {
+      param = paramMap.get(paramName);
+    } else {
+      throw new ExecutorException("Could not find parameter '" + paramName + "'. "
+          + "Note that when there are multiple parameters, 'keyProperty' must include the parameter name (e.g. 'param.id'). "
+          + "Specified key properties are " + ArrayUtil.toString(keyProperties) + " and available parameters are "
+          + paramMap.keySet());
+    }
+    // Remove param name from 'keyProperty' string. e.g. 'param.id' -> 'id'
+    String[] modifiedKeyProperties = new String[keyProperties.length];
+    for (int i = 0; i < keyProperties.length; i++) {
+      if (keyProperties[i].charAt(firstDot) == '.' && keyProperties[i].startsWith(paramName)) {
+        modifiedKeyProperties[i] = keyProperties[i].substring(firstDot + 1);
+      } else {
+        throw new ExecutorException("Assigning generated keys to multiple parameters is not supported. "
+            + "Note that when there are multiple parameters, 'keyProperty' must include the parameter name (e.g. 'param.id'). "
+            + "Specified key properties are " + ArrayUtil.toString(keyProperties) + " and available parameters are "
+            + paramMap.keySet());
+      }
+    }
+    assignKeysToParam(configuration, rs, modifiedKeyProperties, param);
+  }
+
+  private void assignKeysToParam(final Configuration configuration, ResultSet rs, final String[] keyProperties,
+      Object param)
+      throws SQLException {
+    final TypeHandlerRegistry typeHandlerRegistry = configuration.getTypeHandlerRegistry();
+    final ResultSetMetaData rsmd = rs.getMetaData();
+    // Wrap the parameter in Collection to normalize the logic.
+    Collection<?> paramAsCollection = null;
+    if (param instanceof Object[]) {
+      paramAsCollection = Arrays.asList((Object[]) param);
+    } else if (!(param instanceof Collection)) {
+      paramAsCollection = Arrays.asList(param);
+    } else {
+      paramAsCollection = (Collection<?>) param;
+    }
+    TypeHandler<?>[] typeHandlers = null;
+    for (Object obj : paramAsCollection) {
+      if (!rs.next()) {
+        break;
       }
+      MetaObject metaParam = configuration.newMetaObject(obj);
+      if (typeHandlers == null) {
+        typeHandlers = getTypeHandlers(typeHandlerRegistry, metaParam, keyProperties, rsmd);
+      }
+      populateKeys(rs, metaParam, keyProperties, typeHandlers);
     }
-    if (parameters == null) {
-      parameters = new ArrayList<Object>();
-      parameters.add(parameter);
+  }
+
+  private Object getSoleParameter(Object parameter) {
+    if (!(parameter instanceof ParamMap || parameter instanceof StrictMap)) {
```

Comment:
Could it support following code ?
 
```java
List<Country> countries = ...;
SomeBean someBean = ...;

Map<String, Object> params = new HashMap<>();
params.put("list", countries);
params.put("someBean", someBean);

sqlSession.update("...", params);
```

In this case,  is need to use the `DefaultSqlSession$StrictMap` as follows ?

```java
List<Country> countries = ...;
SomeBean someBean = ...;

Map<String, Object> params = new StrictMap<Object>();
params.put("list", countries);
params.put("someBean", someBean);

sqlSession.update("...", params);
```

Reply:
Thank you, @kazuki43zoo .
This PR does not support it (I intentionally dropped it, but forgot about it).
With this PR, when calling SqlSession methods directly (i.e. using XML mapper only), the target list, collection or array must be the only parameter.

Do you think that usage is important?
If so, I would take a different approach i.e. 1) introducing a new syntax `list[].id` for `keyProperty` (this was proposed in the original issue #324) or 2) introducing a new option `keyCollection` to specify the target list, collection or array.

@jeffgbutler , You reviewed the original PR. Any thoughts on the spec?

Reply:
> Do you think that usage is important?

No. However, I prefer to announce dropping backward compatibility and introduce workaround in a release note(or migration guide).

Reply:
Yes, it should be mentioned in the release note.
By workaround, you mean the code using StrictMap? I thought that was an internal class.
The workaround I had in mind was to use Java mapper. Could this be a big limitation?

Reply:
> By workaround, you mean the code using StrictMap? I thought that was an internal class.
The workaround I had in mind was to use Java mapper.

I agree.